### PR TITLE
add summary of x::T output functions

### PR DESCRIPTION
**Pull request by stevengj**
This adds a brief summary of the different output functions (`display`, `show`, `print`, `repr`, etc.) and how they interact to the custom pretty-printing section of the manual, following [this post on discourse](https://discourse.julialang.org/t/how-to-display-a-sparse-array-on-the-repl/113565/5), since people often get confused by this.

### CODE REVIEW

**Comment by halleysfifthinc** [comment]
The `:compact => true` flag isn't set for all container elements, it depends on the shape:

```julia
julia> struct A end; a = A()
A()

julia> function Base.show(io::IO, a::A)
           print(io, get(io, :compact, false))
       end

julia> [ a; a ]
2-element Vector{A}:
 false
 false

julia> [ a a ]
1×2 Matrix{A}:
 true  true
```

Not sure if right here would be the right place, but somewhere it would be good to add the clarification that (afaict) compact is about textual width/columns while limit is about printing textual lines/rows.

**Comment by thofma** [comment]
Regarding the container printing, maybe we can add that it uses the 3-arg version if it does not contain a newline and the 2-arg version otherwise?

File: doc/src/manual/types.md [hunk 1]
```diff
@@ -1587,6 +1587,18 @@ julia> [Polar(3, 4.0) Polar(4.0,5.3)]
 See the [`IOContext`](@ref) documentation for a list of common properties which can be used
 to adjust printing.
 
+### Output-function summary
+
+Here is a brief summary of the different output functions in Julia and how they are related.  Most new types should only need to define `show` methods, if anything.
```
Comment (fingolfin) [inline doc/src/manual/types.md:1592]: Maybe avoid overlong lines, as the rest of the file? I guess < 100 chars or so?
```suggestion
Here is a brief summary of the different output functions in Julia and how they are related.
Most new types should only need to define `show` methods, if anything.
```

File: doc/src/manual/types.md [hunk 2]
```diff
@@ -1587,6 +1587,18 @@ julia> [Polar(3, 4.0) Polar(4.0,5.3)]
...
+
+* [`display(x)`](@ref) tells the current environment to display `x` in whatever way it thinks best. (This might even be a graphical display in something like a Jupyter or Pluto notebook.) By default (e.g. in scripts or in the text REPL), it calls `show(io, "text/plain", x)`, or equivalently `show(io, MIME"text/plain"(), x)`, for an appropriate `io` stream. (In the REPL, `io` is an [`IOContext`](@ref) wrapper around [`stdout`](@ref).) The REPL uses `display` to output the result of an evaluated expression.
+* The 3-argument [`show(io, ::MIME"text/plain", x)`](@ref) method is verbose pretty-printing of `x`. By default (if no 3-argument method is defined for `typeof(x)`), it calls the 2-argument `show(io, x)`.  It is called by the 2-argument `repr("text/plain", x)`.   Other 3-argument `show` methods can be defined for additional MIME types as discussed above, to enable richer display of `x` in some interactive environments.
```
Comment (fingolfin) [inline doc/src/manual/types.md:1595]: "The [...] method is verbose pretty-printing of `x`." seems grammatically dubious. Perhaps replace "is" by "performs" or "does" or so? Depending on what you want to express, of which I am not sure.

File: doc/src/manual/types.md [hunk 3]
```diff
@@ -1587,6 +1587,18 @@ julia> [Polar(3, 4.0) Polar(4.0,5.3)]
...
+* The 2-argument [`show(io, x)`](@ref) is the default simple text representation of `x`. It is called by the 1-argument [`repr(x)`](@ref), and is typically the format you might employ to input `x` into Julia. The 1-argument `show(x)` calls `show(stdout, x)`.
```
Comment (fingolfin) [inline doc/src/manual/types.md:1596]: I am not so sure about that "typically" :-). But OK...

File: doc/src/manual/types.md [hunk 4]
```diff
@@ -1587,6 +1587,18 @@ julia> [Polar(3, 4.0) Polar(4.0,5.3)]
...
+* [`print(io, x)`](@ref) by default calls `show(io, x)`, but a few types have a distinct `print` format — most notably, when `x` is a string, `print` outputs the raw text whereas `show` outputs an escaped string enclosed in quotation marks.  The 1-argument `print(x)` calls `print(stdout, x)`. `print` is also called by [`string(x)`](@ref).
+* [`write(io, x)`](@ref), if it is defined (it generally has *no* default definition for new types), writes a "raw" binary representation of `x` to `io`, e.g. an `x::Int32` will be written as 4 bytes.
+
+It's also helpful to be familiar with the metadata that can be attached to an `io` stream by an [`IOContext`](@ref) wrapper.  For example, the REPL sets the `:limit => true` flag from `display` for an evaluated expression, in order to limit the output to fit in the terminal; you can query this flag with `get(io, :limit, false)`.   And when displaying an object contained within, for example, a multi-column matrix, the `:compact => true` flag could be set, which you can query with `get(io, :compact, false)`.
```
Comment (fingolfin) [inline doc/src/manual/types.md:1600]:
```suggestion
It is also helpful to be familiar with the metadata that can be attached to an `io` stream by an [`IOContext`](@ref) wrapper.  For example, the REPL sets the `:limit => true` flag from `display` for an evaluated expression, in order to limit the output to fit in the terminal; you can query this flag with `get(io, :limit, false)`.   And when displaying an object contained within, for example, a multi-column matrix, the `:compact => true` flag could be set, which you can query with `get(io, :compact, false)`.
```

File: doc/src/manual/types.md (hunk 2 above)
Comment (stevengj) [reply doc/src/manual/types.md:1596]:
```suggestion
* The 3-argument [`show(io, ::MIME"text/plain", x)`](@ref) method performs verbose pretty-printing of `x`. By default (if no 3-argument method is defined for `typeof(x)`), it calls the 2-argument `show(io, x)`.  It is called by the 2-argument `repr("text/plain", x)`.   Other 3-argument `show` methods can be defined for additional MIME types as discussed above, to enable richer display of `x` in some interactive environments.
```

**Comment by stevengj** [comment]
> Regarding the container printing, maybe we can add that it uses the 3-arg version if it does not contain a newline and the 2-arg version otherwise?

I'd rather address that in a separate PR.  I honestly think this behavior (introduced in #34387 as a part of a date/time printing change) should just be reverted; I don't want to bake it into the documentation.

**Pull request merged**